### PR TITLE
Don't publish with unsupported QoS

### PR DESCRIPTION
**Pull request by smurfix**
MQTT 3.3.1.2:
If the Server included a Maximum QoS in its CONNACK response to a Client and it receives a PUBLISH packet with a QoS greater than this, then it uses DISCONNECT with Reason Code 0x9B (QoS not supported) …

### CODE REVIEW

**Comment by agronholm** [comment]
I would be more favorable to a `maximum_qos` property.

**Comment by smurfix** [comment]
Well your wish is my command. *g*

File: src/mqttproto/client_state_machine.py
Comment (agronholm) [inline src/mqttproto/client_state_machine.py:160]: I'm really not sure that silently downgrading the QoS is a good idea. How do other MQTT clients deal with this?

Comment (agronholm) [reply src/mqttproto/client_state_machine.py:160]: I looked at paho-mqtt and it doesn't look like they're doing this. Doing so in the state machine is even worse than in the client implementation. Exposing the `maximum_qos` information was a good call, however.

Comment (agronholm) [reply src/mqttproto/client_state_machine.py:160]: What the state machine could do is raise an error when attempting to send a message with an unsupported QoS.

Comment (smurfix) [reply src/mqttproto/client_state_machine.py:160]: Or maybe we should just drop this line and let the client app deal with the fallout instead, same as with the Retain flag handling.

Comment (agronholm) [reply src/mqttproto/client_state_machine.py:160]: Maybe, yeah, so long as we expose the maximum support QoS, this would seem reasonable.

File: tests/test_async_client.py
Comment (agronholm) [inline tests/test_async_client.py:17]: What do these additions actually test for? This updated test passes just fine without the other changes in this PR.

Comment (smurfix) [reply tests/test_async_client.py:17]: Broker exercising. ;-)

I initially added this code to check whether it was the publish or the subscribe part that caused the problem I was seeing (ultimately fixed in #20).

I left it in because checking every possible QoS combination seems prudent, so that we don't hit any other issues along these lines.

Comment (agronholm) [reply tests/test_async_client.py:17]: But we're not really checking anything new here, are we? When I originally wrote this test, it wasn't passing because the I/O implementation wasn't handling the different QoS levels correctly. If these new test combinations don't add anything useful to the test suite, they should be left out. We're not trying to to validate the brokers here.

Comment (smurfix) [reply tests/test_async_client.py:17]: The original test didn't tell the broker which QoS level we want the messages to have, and the test suite should include that.

Alternately I could split this test in two: one that exercises every `publish` QoS level, and one that does the same with all QoS levels in `subscribe`. That'd result in running six testcases instead of nine, which seems like a somewhat marginal improvement, but it's your call.

Comment (agronholm) [reply tests/test_async_client.py:17]: If the intent is to test if we client can limit the maximum QoS level in its subscription, then the test should at least check the QoS level of the received messages.

Comment (smurfix) [reply tests/test_async_client.py:17]: OK, that's easy enough to add.

File: src/mqttproto/async_client.py
Comment (agronholm) [inline src/mqttproto/async_client.py:488]: Now we just need test coverage for this.

File: src/mqttproto/client_state_machine.py
Comment (agronholm) [inline src/mqttproto/client_state_machine.py:175]: This should also be explicitly covered by the client state machine tests.

**Review by agronholm — CHANGES_REQUESTED** [review]

File: tests/test_async_client.py
Comment (agronholm) [inline tests/test_async_client.py:20]: Why can't you use `pytest.skip()` right now? Under what circumstances would it currently skip the test?

File: tests/test_state_machines.py
Comment (agronholm) [inline tests/test_state_machines.py:113]: Is this line even needed?
```suggestion
```

File: tests/test_state_machines.py
Comment (agronholm) [inline tests/test_state_machines.py:109]:
```suggestion
    """Test that a limited QoS is processed in the client."""
```

File: tests/test_state_machines.py
Comment (smurfix) [reply tests/test_state_machines.py:113]: Probably not in this case. I'll drop it.

File: tests/test_state_machines.py
Comment (agronholm) [inline tests/test_state_machines.py:108]:
```suggestion
def test_client_limit_qos(qos: QoS) -> None:
```

Comment (smurfix) [reply tests/test_state_machines.py:108]: Ah. Sure. *Sigh*

File: tests/test_async_client.py
Comment (smurfix) [reply tests/test_async_client.py:20]: `pytest.skip` raises an exception which gets wrapped into an `ExceptionGroup`. Same problem as in #22 but I didn't want to replicate the solution from there here, only to then have you deal with a merge conflict.

File: tests/test_async_client.py
Comment (agronholm) [inline tests/test_async_client.py:20]:
```suggestion
            return  # TODO add pytest.skip

```

Comment (agronholm) [reply tests/test_async_client.py:20]: Would you mind allowing me to make changes too so I can make these trivial edits without wasting your time on them?

Comment (smurfix) [reply tests/test_async_client.py:20]: Well yes sure, the problem is that github doesn't even ask me whether to allow that. I haven't yet figured out where *that* setting has moved to these days, it used to be right on the create-a-PR page.

Comment (agronholm) [reply tests/test_async_client.py:20]: Is [this](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) helpful?

Comment (smurfix) [reply tests/test_async_client.py:20]: changed

Comment (smurfix) [reply tests/test_async_client.py:20]: > Is [this](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) helpful?

This option isn't even displayed when I look at the PR(s).

Seems like I'll have to ask github support.

**Review by agronholm — APPROVED** [review]

File: tests/test_async_client.py
Comment (smurfix) [reply tests/test_async_client.py:20]: Ah. The reason was that I submitted them from my MoaT org collection instead of my personal one. *Sigh*.

No explanation why that should be so …

**Pull request merged**
